File: aat-android/src/main/java/ch/bailu/aat/views/preferences/AbsSolidView.java
package ch.bailu.aat.views.preferences;

import android.content.Context;

import javax.annotation.Nonnull;

import ch.bailu.aat.util.ui.theme.UiTheme;
import ch.bailu.aat.views.LabelTextView;
import ch.bailu.aat_lib.preferences.AbsSolidType;
import ch.bailu.aat_lib.preferences.OnPreferencesChanged;
import ch.bailu.aat_lib.preferences.StorageInterface;

public abstract class AbsSolidView extends LabelTextView
        implements OnPreferencesChanged {

    private final AbsSolidType solid;

    public AbsSolidView(Context context, final AbsSolidType s, UiTheme theme) {
        super(context, s.getLabel(), theme);
        solid = s;

        theme.button(this);
        setOnClickListener(v -> onRequestNewValue());
    }

    public abstract void onRequestNewValue();

    public void setText() {
        setText("["+solid.toString()+"]");
        setToolTip(solid);
    }

    @Override
    public void onAttachedToWindow() {
        super.onAttachedToWindow();
        solid.register(this);
        setText();
    }

    @Override
    public void onPreferencesChanged(@Nonnull StorageInterface s, @Nonnull String key) {
        if (solid.hasKey(key)) {
            setText();
        }
    }

    @Override
    public void onDetachedFromWindow() {
        super.onDetachedFromWindow();
        solid.unregister(this);
    }
}

File: aat-android/src/main/kotlin/ch/bailu/aat/views/preferences/AbsSolidView.kt
package ch.bailu.aat.views.preferences

import android.content.Context
import ch.bailu.aat.util.ui.theme.UiTheme
import ch.bailu.aat.views.layout.LabelTextView
import ch.bailu.aat_lib.preferences.AbsSolidType
import ch.bailu.aat_lib.preferences.OnPreferencesChanged
import ch.bailu.aat_lib.preferences.StorageInterface

abstract class AbsSolidView(context: Context, private val solid: AbsSolidType, theme: UiTheme) :
    LabelTextView(context, solid.getLabel(), theme), OnPreferencesChanged {
    init {
        theme.button(this)
        setOnClickListener { onRequestNewValue() }
    }

    abstract fun onRequestNewValue()
    fun setText() {
        setText("[$solid]")
        setToolTip(solid)
    }

    public override fun onAttachedToWindow() {
        super.onAttachedToWindow()
        solid.register(this)
        setText()
    }

    override fun onPreferencesChanged(storage: StorageInterface, key: String) {
        if (solid.hasKey(key)) {
            setText()
        }
    }

    public override fun onDetachedFromWindow() {
        super.onDetachedFromWindow()
        solid.unregister(this)
    }
}
